File: java/src/main/java/com/allstask/api/Features.java
// this file is @generated
package com.allstask.api;

import com.allstask.AllstaskHttpClient;
import com.allstask.Utils;
import com.allstask.exceptions.ApiException;
import com.allstask.models.Feature;
import com.allstask.models.FeatureListResponse;

import okhttp3.HttpUrl;

import java.io.IOException;

public class Features {
    private final AllstaskHttpClient client;

    public Features(AllstaskHttpClient client) {
        this.client = client;
    }

    /** */
    public FeatureListResponse listFeatures() throws IOException, ApiException {

        return this.listFeatures(new FeaturesListFeaturesOptions());
    }

    /** */
    public FeatureListResponse listFeatures(final FeaturesListFeaturesOptions options)
            throws IOException, ApiException {
        HttpUrl.Builder url = this.client.newUrlBuilder().encodedPath("/api/v1/features");
        if (options.statuses != null) {
            Utils.addExplodedQueryParameter(url, "statuses", options.statuses);
        }
        if (options.productId != null) {
            url.addQueryParameter("product_id", Utils.serializeQueryParam(options.productId));
        }
        if (options.search != null) {
            url.addQueryParameter("search", options.search);
        }
        if (options.page != null) {
            url.addQueryParameter("page", Utils.serializeQueryParam(options.page));
        }
        if (options.perPage != null) {
            url.addQueryParameter("per_page", Utils.serializeQueryParam(options.perPage));
        }
        return this.client.executeRequest(
                "GET", url.build(), null, null, FeatureListResponse.class);
    }

    /** */
    public Feature getFeature(final String idOrCode) throws IOException, ApiException {
        HttpUrl.Builder url =
                this.client
                        .newUrlBuilder()
                        .encodedPath(String.format("/api/v1/features/%s", idOrCode));
        return this.client.executeRequest("GET", url.build(), null, null, Feature.class);
    }
}

File: typescript/src/api/features.ts
// this file is @generated
import {
    Feature,
    FeatureSerializer,
} from '../models/feature';
import {
    FeatureListResponse,
    FeatureListResponseSerializer,
} from '../models/featureListResponse';
import {
    FeatureStatus,
    FeatureStatusSerializer,
} from '../models/featureStatus';
import {
    ProductId,
    ProductIdSerializer,
} from '../models/productId';
import { HttpMethod, AllstaskRequest, AllstaskRequestContext } from "../request";

export interface FeaturesListFeaturesOptions {
        /** Filter by feature status. Repeat the param to select multiple, omit to return all. */
            statuses?: FeatureStatus[];
        /** Filter by product. Omit to return features across all products. */
            productId?: ProductId;
        /** Search by feature name. */
            search?: string;
        /** Page number (0-indexed) */
            page?: number;
        /** Number of items per page */
            perPage?: number;
        }

    export class Features {
    public constructor(private readonly requestCtx: AllstaskRequestContext) {}

    /**  */
        public listFeatures(
            options?: FeaturesListFeaturesOptions,
            ): Promise<FeatureListResponse> {
            const request = new AllstaskRequest(HttpMethod.GET, "/api/v1/features");

            request.setQueryParam("statuses", options?.statuses);
            request.setQueryParam("product_id", options?.productId);
            request.setQueryParam("search", options?.search);
            request.setQueryParam("page", options?.page);
            request.setQueryParam("per_page", options?.perPage);
            
                return request.send(
                    this.requestCtx,
                    FeatureListResponseSerializer._fromJsonObject,
                );
            }

        

    /**  */
        public getFeature(
            idOrCode: string,
            ): Promise<Feature> {
            const request = new AllstaskRequest(HttpMethod.GET, "/api/v1/features/{id_or_code}");

            request.setPathParam("id_or_code", idOrCode);
            
                return request.send(
                    this.requestCtx,
                    FeatureSerializer._fromJsonObject,
                );
            }

        

    }

